mev-build-rs/src/payload/builder.rs

```diff
@@ -200,26 +212,66 @@ impl Deref for PayloadBuilder {
 #[derive(Debug)]
 pub struct Inner {
     bids: Sender<EthBuiltPayload>,
-    signer: LocalWallet,
-    pub fee_recipient: Address,
+    signer: PrivateKeySigner,
+    fee_recipient: Address,
     chain_id: ChainId,
-    states: Mutex<HashMap<PayloadId, BundleStateWithReceipts>>,
+    execution_outcomes: Mutex<HashMap<PayloadId, ExecutionOutcome>>,
+    evm_config: EthEvmConfig,
 }
 
 impl PayloadBuilder {
     pub fn new(
         bids: Sender<EthBuiltPayload>,
-        signer: LocalWallet,
+        signer: PrivateKeySigner,
         fee_recipient: Address,
         chain_id: ChainId,
+        chain_spec: Arc<ChainSpec>,
     ) -> Self {
-        let inner = Inner { bids, signer, fee_recipient, chain_id, states: Default::default() };
+        let evm_config = EthEvmConfig::new(chain_spec);
+        let inner = Inner {
+            bids,
+            signer,
+            fee_recipient,
+            chain_id,
+            execution_outcomes: Default::default(),
+            evm_config,
+        };
         Self(Arc::new(inner))
     }
 
-    pub fn get_build_state(&self, payload_id: PayloadId) -> Option<BundleStateWithReceipts> {
-        let mut state = self.states.lock().expect("can lock");
-        state.remove(&payload_id)
+    pub fn cfg_and_block_env(
+        &self,
+        payload_config: &PayloadConfig<BuilderPayloadBuilderAttributes>,
+    ) -> (CfgEnvWithHandlerCfg, BlockEnv) {
+        let next_attributes = NextBlockEnvAttributes {
+            timestamp: payload_config.attributes.timestamp(),
+            suggested_fee_recipient: payload_config.attributes.suggested_fee_recipient(),
+            prev_randao: payload_config.attributes.prev_randao(),
+        };
+        let (cfg_env, mut block_env) = self
+            .evm_config
+            .next_cfg_and_block_env(payload_config.parent_block.header.header(), next_attributes);
+
+        // if there is a proposal attributes present, then set the gas limit and fee recipient
+        if let Some(ref proposal_attributes) = payload_config.attributes.proposal {
+            let gas_limit = compute_preferred_gas_limit(
+                proposal_attributes.proposer_gas_limit,
+                payload_config.parent_block.gas_limit,
+            );
+            // NOTE: reserve enough gas for the final payment transaction
+            block_env.gas_limit = U256::from(gas_limit) - U256::from(BASE_TX_GAS_LIMIT);
+
+            block_env.coinbase = proposal_attributes.proposer_fee_recipient;
+        } else {
+            block_env.coinbase = self.0.fee_recipient;
+        }
```

Comment:
pretty sure we always want this to be the `self.0.fee_recipient`, aka the builder's fee recipient

if we set coinbase to the proposer's fee recipient, then the builder gets 0 value and on top of that pays the proposer an additional amount